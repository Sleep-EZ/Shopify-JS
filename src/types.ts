export const SHOPIFY_TYPE_PRODUCT = 'product';
export const SHOPIFY_TYPE_COLLECTION = 'collection';
export const SHOPIFY_TYPE_PAGE = 'page';

export type ShopifyTypeStr = (
    typeof SHOPIFY_TYPE_PRODUCT |
    typeof SHOPIFY_TYPE_COLLECTION |
    typeof SHOPIFY_TYPE_PAGE);

export type ShopifyType<H extends string> = (
    Product<H> | Collection<H> | Page<H>);

type Expires = { __ts: number };

/**
 * Type definitions for the core Shopify JSON representations.
 *
 * These definitions are based on Shopify's web JSON API,
 * and certainly has the possibility for change (without notice).
 */
export type Product<H> = Expires & {
    id: number,
    handle: H,
    body_html: string | null,
    vendor: string | null,
    product_type: string | null,
    created_at: Date,
    updated_at: Date,
    published_at: Date,
    template_suffix: string,
    published_scope: string,
    tags: string[],
    variants: Variant[],
    options: Option[],
    images: Image[],
    image: Image,
};

export type Image = {
    id: number,
    product_id: number,
    position: number,
    created_at: Date,
    updated_at: Date,
    width: number,
    height: number,
    src: string,
    variant_ids: number[],
};

export type Option = {
    id: number,
    product_id: number,
    name: string,
    position: number,
    values: string[],
};

export type Variant = {
    id: number,
    product_id: number,
    title: string,
    price: number,
    sku: string,
    position: number,
    inventory_policy: string,
    compare_at_price: number,
    fulfillment_service: string,
    inventory_management: string | null,
    option1: string,
    option2: string | null,
    option3: string | null,
    created_at: Date,
    updated_at: Date,
    taxable: boolean,
    barcode: string | null,
    grams: number,
    image_id: number,
    inventory_quantity: number,
    weight: number,
    weight_unit: string,
    inventory_item_id: number,
    old_inventory_quantity: number,
    requires_shipping: boolean,
};

export type Page<H> = Expires & {
    id: number,
    title: string, 
    handle: H,

    body_html: string,
    created_at: Date,
    published_at: Date,
    updated_at: Date,
};

export type Collection<H> = Expires & {
    id: number,
    title: string,
    handle: H,
    description: string,
    published_at: Date,
    updated_at: Date,
    image: Image,
    products_count: number,

    // This is a field provided by this API only
    products: Product<string>[],
}