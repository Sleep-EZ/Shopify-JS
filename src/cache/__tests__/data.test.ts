import {testCreatePage, testCreateProduct, testCreateVariant} from '../../testing/index';
import {ShopifyTypeEnum} from '../../types';
import {Cache, CacheData$Values, generateEmptyCacheData, indexSingleElement, rebuildCache} from '../index';

describe('CacheData', () => {
  it('generateEmptyCacheData() is unchanged', () => {
    expect(generateEmptyCacheData()).toMatchSnapshot();
  });

  describe('indexSingleElement()', () => {
    it('returns false when item already exists', () => {
      const cache = new Cache();
      const testPage = testCreatePage({
        id: 1,
        handle: 'test',
      });

      cache._cache.ids[1] = 0;
      cache._cache.handles[ShopifyTypeEnum.Page]['test'] = 0;

      expect(indexSingleElement(cache._cache, testPage)).toBeFalsy();
    });

    it('covpath: indexing a variant skips indexing a handle', () => {
      const cache = new Cache();
      const variant = testCreateVariant({id: 1});

      indexSingleElement(cache._cache, variant);
      expect(cache._cache.ids).toEqual({1: 0});
      expect(cache._cache.data).toEqual([variant]);
    });

    it('covpath: re-indexing a variant overwrites the value', () => {
      const cache = new Cache();
      const product = testCreateProduct({id: 1});
      const variant = testCreateVariant({id: 10});
      const modifiedVariant = Object.assign(variant, {title: 'A'});

      product.variants = [variant];
      indexSingleElement(cache._cache, product);

      product.variants = [modifiedVariant];
      indexSingleElement(cache._cache, product);

      expect(cache._cache.ids).toEqual({1: 0, 10: 1});
      expect(cache._cache.data).toEqual([product, modifiedVariant]);
    });

    it('also indexes product variants', () => {
      const cache = new Cache();
      const variant1 = testCreateVariant({id: 10});
      const product = testCreateProduct({
        id: 1,
        handle: 'example',
        variants: [variant1],
      });

      indexSingleElement(cache._cache, product);

      expect(Object.keys(cache._cache.ids)).toEqual(['1', '10']);
      expect(cache._cache.data).toEqual([product, variant1]);
    });

    it('SPECIAL: Indexing handle resolves proper ID', () => {
      const cache = new Cache();
      const testPage = testCreatePage({id: 1, handle: 'sad'});
      const testPage2 = testCreatePage({id: 2, handle: 'happy'});

      cache._cache.ids[1] = 0;
      cache._cache.ids[2] = 1;
      cache._cache.handles[ShopifyTypeEnum.Page]['happy'] = 1;
      // 'sad' is missing in the index

      // Should still be false, nothing was created, only an index was added
      expect(indexSingleElement(cache._cache, testPage)).toBeFalsy();
      expect(indexSingleElement(cache._cache, testPage2)).toBeFalsy();
      expect(cache._cache.handles[ShopifyTypeEnum.Page])
          .toEqual({happy: 1, sad: 0});
    });
  });

  describe('rebuildCache', () => {
    const page1 = testCreatePage({id: 1, handle: 'page1'});
    const page2 = testCreatePage({id: 2, handle: 'page2'});

    const uncleanCache: CacheData$Values = [page1, null, page2];
    const rebuiltCache = rebuildCache(uncleanCache);

    expect(rebuiltCache.data).toEqual([page1, page2]);
    expect(rebuiltCache.data !== uncleanCache).toBeTruthy();
  });
});