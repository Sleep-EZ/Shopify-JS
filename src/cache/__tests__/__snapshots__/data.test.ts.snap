// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CacheData generateEmptyCacheData() is unchanged 1`] = `
Object {
  "data": Array [],
  "handles": Object {
    "collection": Object {},
    "page": Object {},
    "product": Object {},
    "variant": Object {},
  },
  "ids": Object {},
}
`;
